interfaces/builtin/compatibility.go

```diff
@@ -0,0 +1,172 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// CompatRange is a range of unsigned integers, with Min <= Max.
+type CompatRange struct {
+	Min uint
+	Max uint
+}
+
+// CompatTopic represents a <tag>[-<integer|integer range>] (topic) in a compatibility field.
+type CompatTopic struct {
+	// Tag is the topic identifier
+	Tag string
+	// Values is the list of integer/integer ranges. Only the last range
+	// can have Min!=Max.
+	Values []CompatRange
+}
+
+type CompatField struct {
+	Topics []CompatTopic
+}
+
+const (
+	maxTopics   = 3
+	maxIntegers = 3
+)
+
+var (
+	tagRegexp      = regexp.MustCompile(`^[a-z][a-z0-9]{0,31}$`)
+	intRegexp      = regexp.MustCompile(`^[0-9]{1,8}$`)
+	intRangeRegexp = regexp.MustCompile(`^\(([0-9]{1,8})\.\.([0-9]{1,8})\)$`)
+)
+
+func isRealCompatRange(cr CompatRange) bool {
+	return cr.Min < cr.Max
+}
+
+func appendTopic(topics []CompatTopic, topic *CompatTopic) []CompatTopic {
+	if topic != nil {
+		if len(topic.Values) == 0 {
+			// If we have just a tag, that is equivalent to <tag>-0
+			topic.Values = []CompatRange{{0, 0}}
+		}
+		topics = append(topics, *topic)
+	}
+	return topics
+}
+
+func appendRange(ranges []CompatRange, rg CompatRange) ([]CompatRange, error) {
+	numVal := len(ranges)
+	if numVal > 0 && isRealCompatRange(ranges[numVal-1]) {
+		return nil, fmt.Errorf("ranges only allowed at the end of compatibility field")
+	}
+	if numVal == maxIntegers {
+		return nil, fmt.Errorf("only %d integer/integer ranges allowed per topic in compatibility field", maxIntegers)
+	}
+	return append(ranges, rg), nil
+}
+
+// decodeCompatField decodes a compatibility string, which is a sequence of
+// topics separated by dashes. Each topic consist of an alphanumerical
+// descriptor followed by a dash-separated series of integer or integer ranges.
+func decodeCompatField(compat string) (*CompatField, error) {
```

Comment:
should this take the maxTopics, maxIntegers as option perhaps do have the choice to do different things for pack vs runtime?  also shouldn't these helpers live in interfaces and be exported?